### ISSUE TITLE
Lint suggestion: nested format!

### ISSUE THREAD

**Issue by jplatte** [rust-lang/rust-clippy#7667]
### What it does

Warn on `format!` within the arguments of another macro that does formatting such as `format!` itself, `write!` or `println!`.

Suggests replacing the inner `format!` call with `format_args!` or inlining it entirely (see example).

### Categories (optional)

- Kind: style / perf

The recommended code is both shorter and avoids a temporary allocation.

### Drawbacks

There could be situations when the outer macro is `format_args!` where lifetimes don't work out with the `format!` changed to `format_args!` or inlined entirely. I'm not sure how hard it would be to detect these cases.

### Example

```rust
println!("error: {}", format!("something failed at {}", Location::caller()));
```

Could be written as:

```rust
println!("error: {}", format_args!("something failed at {}", Location::caller()));
```

or inlined entirely (when the argument is used exactly once like here):

```rust
println!("error: something failed at {}", Location::caller());
```

**Comment by matthiaskrgr** [comment]
Related issues: https://github.com/rust-lang/rust-clippy/issues/3155 https://github.com/rust-lang/rust-clippy/issues/6259

**Comment by camsteffen** [comment]
Lint name idea: `format_in_format_args`. Note that `panic!(.., format!(..))` is expanded to `..format_args!(.., format!(..))..`. The same is true for all format_args-like macros.

To implement, I would start with `FormatArgsExpn::parse`, then try `FormatExpn::parse` for each value arg, then look at the outer macro expansion to look for `panic!`, `assert!`, `write!`, etc.

**Comment by smoelius** [comment]
@camsteffen If I were to implement this, would it be too invasive if I were to change `FormatExpn` to `FormatLikeExpn` so that it could handle things like `println!`.

Currently, it strictly handles `format!`: https://github.com/rust-lang/rust-clippy/blob/fb17671eb07fc9f68c17b7494e402591fe08b299/clippy_utils/src/higher.rs#L488

**Comment by camsteffen** [comment]
@smoelius I think I would add something like this, on top of what we already have:

```rust
enum FormatLikeExpn {
    Assert(AssertExpn),
    Format(FormatExpn),
    FormatArgs(FormatArgsExpn),
    Print(PrintExpn),
}
```

so we still have the option of parsing a specific macro.

**Comment by smoelius** [comment]
OK. Thanks for your response. I'm going to play with this and see what makes sense.
